From: Chad Monroe <chad@monroe.io>
Date: Wed, 14 Jan 2026 16:02:51 -0800
Subject: [PATCH] net: phy: realtek: add support for RTL8261C/D-CG 10Gbps PHY

The RTL8261C/D-CG is a Realtek 10Gbps copper PHY supporting 100M/1G/2.5G/5G/10G
speeds via USXGMII interface. Requires firmware patching at init via patch
mode entry/exit sequence. Work around MMD7 AN STAT1 not reporting
autonegotiation capability correctly by forcing the Autoneg bit in
get_features.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/phy/realtek/phy_patch_rtl826x.c |   53 +++++
 drivers/net/phy/realtek/phy_patch_rtl826x.h |    1 
 drivers/net/phy/realtek/realtek_main.c      |  187 ++++++++++++++++++
 3 files changed, 241 insertions(+)

--- a/drivers/net/phy/realtek/phy_patch_rtl826x.c
+++ b/drivers/net/phy/realtek/phy_patch_rtl826x.c
@@ -214,6 +214,60 @@ static int phy_patch_rtl826x_op(struct p
 	return ret;
 }
 
+int phy_patch_rtl8261c_db_init(struct phy_device *phydev)
+{
+	int ret = 0, i;
+	struct device *dev = &phydev->mdio.dev;
+	struct rtl826x_priv *priv = phydev->priv;
+	struct rt_phy_patch_db *patch_db = NULL;
+	const struct firmware *fw = NULL;
+	struct rtk_hwpatch *patches;
+
+	ret = request_firmware_direct(&fw, "rtl8261c.bin", dev);
+	if (ret) {
+		phydev_err(phydev, "Failed to request rtl8261c.bin\n");
+		goto out;
+	}
+	patches = (struct rtk_hwpatch *)fw->data;
+
+	patch_db = kmalloc(sizeof(struct rt_phy_patch_db), GFP_ATOMIC);
+	if (!patch_db) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	memset(patch_db, 0x0, sizeof(struct rt_phy_patch_db));
+
+	patch_db->patch_op = phy_patch_rtl826x_op;
+	patch_db->size = fw->size;
+
+	patch_db->conf = kmalloc(patch_db->size, GFP_ATOMIC);
+	if (!patch_db->conf) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	for (i = 0; i < fw->size / sizeof(struct rtk_hwpatch); i++) {
+		patch_db->conf[i].patch_op = patches[i].patch_op;
+		patch_db->conf[i].portmask = patches[i].portmask;
+		patch_db->conf[i].pagemmd = le16_to_cpu(patches[i].pagemmd);
+		patch_db->conf[i].addr = le16_to_cpu(patches[i].addr);
+		patch_db->conf[i].msb = patches[i].msb;
+		patch_db->conf[i].lsb = patches[i].lsb;
+		patch_db->conf[i].data = le16_to_cpu(patches[i].data);
+	}
+
+	priv->patch = patch_db;
+
+out:
+	if (ret)
+		kfree(patch_db);
+	if (fw && !IS_ERR(fw))
+		release_firmware(fw);
+
+	return ret;
+}
+
 int phy_patch_rtl8261n_db_init(struct phy_device *phydev)
 {
 	int ret = 0, i;
@@ -360,5 +414,6 @@ out:
 	return ret;
 }
 
+MODULE_FIRMWARE("rtl8261c.bin");
 MODULE_FIRMWARE("rtl8261n.bin");
-MODULE_FIRMWARE("rtl8264b.bin");
\ No newline at end of file
+MODULE_FIRMWARE("rtl8264b.bin");
--- a/drivers/net/phy/realtek/phy_patch_rtl826x.h
+++ b/drivers/net/phy/realtek/phy_patch_rtl826x.h
@@ -3,6 +3,7 @@
 
 #include "phy_patch.h"
 
+int phy_patch_rtl8261c_db_init(struct phy_device *phydev);
 int phy_patch_rtl8261n_db_init(struct phy_device *phydev);
 int phy_patch_rtl8264b_db_init(struct phy_device *phydev);
 
--- a/drivers/net/phy/realtek/realtek_main.c
+++ b/drivers/net/phy/realtek/realtek_main.c
@@ -163,6 +163,8 @@
 
 #define RTL8224_SRAM_RTCT_FAULT(pair)		(0x8026 + (pair) * 4)
 #define RTL8224_SRAM_RTCT_FAULT_BUSY		BIT(0)
+#define RTL8224_SRAM_RTCT_FAULT_MISMATCH_OPEN	BIT(1)
+#define RTL8224_SRAM_RTCT_FAULT_MISMATCH_SHORT	BIT(2)
 #define RTL8224_SRAM_RTCT_FAULT_OPEN		BIT(3)
 #define RTL8224_SRAM_RTCT_FAULT_SAME_SHORT	BIT(4)
 #define RTL8224_SRAM_RTCT_FAULT_OK		BIT(5)
@@ -225,6 +227,8 @@
 #define PHYID_RTL8224				0x001ccad0
 #define PHYID_RTL8264				0x001ccaf2
 #define PHYID_RTL8261N				0x001ccaf3
+#define PHYID_RTL8261C_CG			0x001cc898
+#define PHYID_RTL8261D_CG			0x001cc898
 #define PHYID_RTL9000A				0x001ccb00
 
 /* RTL8211E and RTL8211F support up to three LEDs */
@@ -240,6 +244,33 @@
 
 #define RTL826X_VND2_INSR				0xA43A
 
+/* RTL8261C-CG specific registers */
+#define RTL8261C_VND2_PHY_STATE			0xA420
+#define RTL8261C_VND2_PHY_STATE_READY_MASK	0x0003
+#define RTL8261C_VND2_PHY_STATE_READY		0x0003
+
+#define RTL8261C_VND2_PATCH_MODE_CTRL		0xB820
+#define RTL8261C_VND2_PATCH_MODE_STAT		0xB800
+#define RTL8261C_VND2_PATCH_CTRL		0xB82E
+
+#define RTL8261C_VND2_INDIRECT_ADDR		0xA436
+#define RTL8261C_VND2_INDIRECT_DATA		0xA438
+
+#define RTL8261C_VND1_RESET_CTRL		0x145
+
+#define RTL8261C_VND2_THERMAL_SENSOR_CTRL	0xB54C
+#define RTL8261C_THERMAL_SENSOR_ENABLE		BIT(3)
+
+/* RTL8261C-CG indirect register addresses */
+#define RTL8261C_INDIRECT_PATCH_VER		0x801E
+#define RTL8261C_INDIRECT_PATCH_SETUP		0x8023
+
+/* RTL8261C-CG patch mode values */
+#define RTL8261C_PATCH_MODE_ENTER		BIT(4)
+#define RTL8261C_PATCH_MODE_READY		BIT(6)
+#define RTL8261C_PATCH_MODE_SETUP		0x0090
+#define RTL8261C_PATCH_SETUP_VAL		0x6100
+
 MODULE_DESCRIPTION("Realtek PHY driver");
 MODULE_AUTHOR("Johnson Leung");
 MODULE_LICENSE("GPL");
@@ -1824,6 +1855,10 @@ static int rtl8224_cable_test_result_tra
 	if (result & RTL8224_SRAM_RTCT_FAULT_CROSS_SHORT)
 		return ETHTOOL_A_CABLE_RESULT_CODE_CROSS_SHORT;
 
+	if (result & (RTL8224_SRAM_RTCT_FAULT_MISMATCH_OPEN |
+		      RTL8224_SRAM_RTCT_FAULT_MISMATCH_SHORT))
+		return ETHTOOL_A_CABLE_RESULT_CODE_IMPEDANCE_MISMATCH;
+
 	return ETHTOOL_A_CABLE_RESULT_CODE_UNSPEC;
 }
 
@@ -2416,6 +2451,121 @@ static int rtl8264b_probe(struct phy_dev
 	return 0;
 }
 
+static int rtl8261c_cg_match_phy_device(struct phy_device *phydev,
+					const struct phy_driver *phydrv)
+{
+	return rtlgen_is_c45_match(phydev, PHYID_RTL8261C_CG, true);
+}
+
+static int rtl8261c_cg_probe(struct phy_device *phydev)
+{
+	int ret;
+
+	ret = rtl826x_probe(phydev);
+	if (ret < 0)
+		return ret;
+
+	ret = phy_patch_rtl8261c_db_init(phydev);
+	if (ret < 0)
+		return -EPROBE_DEFER;
+
+	phy_set_bits_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_THERMAL_SENSOR_CTRL,
+			 RTL8261C_THERMAL_SENSOR_ENABLE);
+
+	return 0;
+}
+
+static int rtl8261c_cg_read_indirect(struct phy_device *phydev, u16 reg)
+{
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_INDIRECT_ADDR, reg);
+	return phy_read_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_INDIRECT_DATA);
+}
+
+static void rtl8261c_cg_write_indirect(struct phy_device *phydev, u16 reg, u16 val)
+{
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_INDIRECT_ADDR, reg);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_INDIRECT_DATA, val);
+}
+
+static int rtl8261c_cg_config_init(struct phy_device *phydev)
+{
+	struct rtl826x_priv *priv = phydev->priv;
+	int ret, val;
+
+	ret = phy_read_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_PHY_STATE);
+	if ((ret & RTL8261C_VND2_PHY_STATE_READY_MASK) == 0) {
+		phy_set_bits_mmd(phydev, MDIO_MMD_VEND1,
+				 RTL8261C_VND1_RESET_CTRL, BIT(0));
+		phy_clear_bits_mmd(phydev, MDIO_MMD_VEND1,
+				   RTL8261C_VND1_RESET_CTRL, BIT(0));
+		mdelay(100);
+	}
+
+	ret = phy_read_mmd_poll_timeout(phydev, MDIO_MMD_VEND2,
+					RTL8261C_VND2_PHY_STATE, val,
+					(val & RTL8261C_VND2_PHY_STATE_READY_MASK) ==
+					RTL8261C_VND2_PHY_STATE_READY,
+					10000, 1000000, false);
+	if (ret)
+		return ret;
+
+	val = rtl8261c_cg_read_indirect(phydev, RTL8261C_INDIRECT_PATCH_VER);
+	if (val == 0x1515)
+		goto configure;
+
+	ret = phy_set_bits_mmd(phydev, MDIO_MMD_VEND2,
+			       RTL8261C_VND2_PATCH_MODE_CTRL,
+			       RTL8261C_PATCH_MODE_ENTER);
+	if (ret < 0)
+		return ret;
+
+	ret = phy_read_mmd_poll_timeout(phydev, MDIO_MMD_VEND2,
+					RTL8261C_VND2_PATCH_MODE_STAT, val,
+					val & RTL8261C_PATCH_MODE_READY,
+					10000, 1000000, false);
+	if (ret)
+		return ret;
+
+	rtl8261c_cg_write_indirect(phydev, RTL8261C_INDIRECT_PATCH_SETUP,
+				   RTL8261C_PATCH_SETUP_VAL);
+	rtl8261c_cg_write_indirect(phydev, RTL8261C_VND2_PATCH_CTRL, 0x0001);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, RTL8261C_VND2_PATCH_MODE_CTRL,
+		      RTL8261C_PATCH_MODE_SETUP);
+
+	ret = phy_patch(phydev);
+	if (ret)
+		return ret;
+
+	phy_clear_bits_mmd(phydev, MDIO_MMD_VEND2,
+			   RTL8261C_VND2_PATCH_CTRL, BIT(0));
+	rtl8261c_cg_write_indirect(phydev, RTL8261C_INDIRECT_PATCH_SETUP, 0x0000);
+
+	phy_clear_bits_mmd(phydev, MDIO_MMD_VEND2,
+			   RTL8261C_VND2_PATCH_MODE_CTRL,
+			   RTL8261C_PATCH_MODE_ENTER);
+
+	ret = phy_read_mmd_poll_timeout(phydev, MDIO_MMD_VEND2,
+					RTL8261C_VND2_PATCH_MODE_STAT, val,
+					!(val & RTL8261C_PATCH_MODE_READY),
+					10000, 1000000, false);
+	if (ret)
+		return ret;
+
+	rtl8261c_cg_write_indirect(phydev, RTL8261C_INDIRECT_PATCH_VER, 0x1515);
+
+configure:
+	if (priv->pnswap_tx)
+		phy_set_bits_mmd(phydev, MDIO_MMD_VEND1,
+				 RTL826X_VEND1_SERDES_GLOBAL_CFG,
+				 RTL826X_VEND1_SERDES_GLOBAL_CFG_HSO_INV);
+	if (priv->pnswap_rx)
+		phy_set_bits_mmd(phydev, MDIO_MMD_VEND1,
+				 RTL826X_VEND1_SERDES_GLOBAL_CFG,
+				 RTL826X_VEND1_SERDES_GLOBAL_CFG_HSI_INV);
+
+	return 0;
+}
+
 static int rtl826x_config_init(struct phy_device *phydev)
 {
 	struct rtl826x_priv *priv = phydev->priv;
@@ -2459,6 +2609,24 @@ static int rtl826x_get_features(struct p
 	return 0;
 }
 
+static int rtl8261c_cg_get_features(struct phy_device *phydev)
+{
+	int ret;
+
+	ret = rtl826x_get_features(phydev);
+	if (ret)
+		return ret;
+
+	/*
+	 * 10GBASE-T requires autonegotiation. the RTL8261C MMD7 AN STAT1 register does
+	 * not report AN capability correctly, causing genphy_c45_pma_read_abilities()
+	 * to not set the Autoneg bit in supported. force it here.
+	 */
+	linkmode_set_bit(ETHTOOL_LINK_MODE_Autoneg_BIT, phydev->supported);
+
+	return 0;
+}
+
 static int rtl825xb_get_features(struct phy_device *phydev)
 {
 	int ret;
@@ -3143,6 +3311,25 @@ static struct phy_driver realtek_drvs[]
 		.get_tunable    = rtl826x_get_tunable,
 		.set_tunable    = rtl826x_set_tunable,
 	}, {
+		.name           = "RTL8261C/D-CG 10Gbps PHY",
+		.config_init    = rtl8261c_cg_config_init,
+		.probe          = rtl8261c_cg_probe,
+		.get_features   = rtl8261c_cg_get_features,
+		.suspend        = genphy_c45_pma_suspend,
+		.resume         = rtlgen_c45_resume,
+		.config_aneg    = rtl826x_config_aneg,
+		.aneg_done      = genphy_c45_aneg_done,
+		.read_status    = rtl826x_read_status,
+		.config_intr    = rtl826x_config_intr,
+		.handle_interrupt = rtl826x_handle_intr,
+		.match_phy_device = rtl8261c_cg_match_phy_device,
+		.set_wol        = rtl826x_set_wol,
+		.get_wol        = rtl826x_get_wol,
+		.get_tunable    = rtl826x_get_tunable,
+		.set_tunable    = rtl826x_set_tunable,
+		.cable_test_start = rtl8224_cable_test_start,
+		.cable_test_get_status = rtl8224_cable_test_get_status,
+	}, {
 		PHY_ID_MATCH_EXACT(PHYID_RTL8264),
 		.name           = "RTL8264 10Gbps PHY",
 		.config_init    = rtl826x_config_init,
